File: packages/markdown/src/front-matter.ts
/** The block of metadata a note may open with, and the surgery on one key of it.
 *
 *  YAML between two `---` fences, which is Obsidian's convention and so this
 *  app's: a note that says what it is says it there, and every other editor that
 *  reads markdown leaves the block alone. What the app keeps in it today is the
 *  title, the author, the language and the date an export reads, the `tags:` a
 *  search reads, and the `icon:` a row in the file list wears.
 *
 *  Reading a key is one line of prose; changing one is not, which is why this is
 *  a module. A note is somebody's file: setting a key has to leave every other
 *  line of the block exactly as it was, taking one away has to leave no empty
 *  block behind, and the change has to come out as an edit of the few characters
 *  that moved rather than as a rewritten note - because a note open in a pane
 *  takes the edit and keeps every caret in it where its reader left it.
 *
 *  Top-level keys only. A key indented under another one belongs to that one:
 *  `paper` under `export:` is not the note's paper, and nothing here pretends
 *  otherwise. */

import { oneEdit, type TextEdit } from './edits'
import { flowItems, listItem, unquoted } from './yaml'

/** Where a note's front matter sits.
 *
 *  All four offsets, because the four questions differ: what the block says is
 *  `body`, what comes after it starts at `to`, a new key goes in at `close`, and
 *  taking the block away is `from` to `to`. */
export interface FrontMatterBlock {
  /** The start of the note, which is the only place a block can open. */
  from: number
  /** Past the closing fence and its line break: where the note's words start. */
  to: number
  /** The lines between the fences, the last one's break included. */
  body: { from: number; to: number }
  /** Where the closing fence's own line begins. */
  close: number
}

/** One key's line, as it is written. `value` is what follows the colon, so a key
 *  with nothing after it has an empty one. */
interface KeyLine {
  /** The whole line, its break not included. */
  from: number
  to: number
  /** The value alone: from just past the colon to the end of the line. */
  value: { from: number; to: number }
}

/** A top-level `key:` and whatever follows it on the line. Indented lines are
 *  somebody else's key, and a line that is only a value is part of a list. */
const KEY = /^([A-Za-z_][\w-]*)[ \t]*:/

/** Where the front matter block sits, or null when the note opens with anything
 *  else - which is most notes.
 *
 *  The opening fence has to be the note's first line and the closing one a line
 *  that says nothing but `---`, so a horizontal rule in the middle of a note
 *  cannot be read as the end of a block that never started. */
export function frontMatterBlock(source: string): FrontMatterBlock | null {
  if (!source.startsWith('---')) return null

  const first = source.indexOf('\n')
  if (first === -1 || source.slice(0, first).trim() !== '---') return null

  let at = first + 1
  while (at <= source.length) {
    const end = source.indexOf('\n', at)
    const stop = end === -1 ? source.length : end

    if (source.slice(at, stop).trim() === '---') {
      return {
        from: 0,
        to: end === -1 ? source.length : end + 1,
        body: { from: first + 1, to: at },
        close: at,
      }
    }

    if (end === -1) break
    at = end + 1
  }

  return null
}

/** The block's own lines as text, or null where there is no block. */
export function frontMatter(source: string): string | null {
  const block = frontMatterBlock(source)
  if (!block) return null

  return source.slice(block.body.from, block.body.to).replace(/\r?\n$/, '')
}

/** The note without its front matter, which is what a renderer is handed:
 *  metadata is not content. */
export function stripFrontMatter(source: string): string {
  const block = frontMatterBlock(source)
  return block ? source.slice(block.to) : source
}

/** A top-level key read as a list.
 *
 *  YAML writes one three ways and a note may use any of them:
 *
 *      aliases: [One, Two]
 *      aliases:
 *        - One
 *        - Two
 *      aliases: One
 *
 *  The last is a list of one, because a reader asking for a list wants the
 *  answer in one shape. Empty where the note has no block, no such key, or
 *  nothing under it. Still not a YAML parser: what it reads is what a note
 *  written by hand or by Obsidian actually holds. */
export function frontMatterList(source: string, key: string): string[] {
  const block = frontMatterBlock(source)
  if (!block) return []

  const line = keyLine(source, block, key)
  if (!line) return []

  const value = source.slice(line.value.from, line.value.to).trim()

  // On the line: a flow sequence, or a single value standing for a list of one.
  if (value) return (flowItems(value) ?? [unquoted(value)]).filter(Boolean)

  // Under it: the `- item` lines, up to the next key of the note's own.
  const out: string[] = []
  let at = line.to + 1

  while (at < block.body.to) {
    const end = source.indexOf('\n', at)
    const stop = end === -1 || end > block.body.to ? block.body.to : end
    const text = source.slice(at, stop)

    if (text.trim()) {
      const item = listItem(text)
      if (item === null) break
      out.push(item)
    }

    if (end === -1) break
    at = end + 1
  }

  return out.filter(Boolean)
}

/** A top-level `key: value` from the front matter, quotes stripped, or null when
 *  the note has no block, no such key, or nothing after the colon. */
export function frontMatterValue(source: string, key: string): string | null {
  const block = frontMatterBlock(source)
  if (!block) return null

  const line = keyLine(source, block, key)
  if (!line) return null

  const value = source.slice(line.value.from, line.value.to).trim()
  return unquoted(value) || null
}

/** The line a key is written on, or null when the block has no such key. The
 *  first one wins, the way YAML reads a duplicated key. */
function keyLine(source: string, block: FrontMatterBlock, key: string): KeyLine | null {
  const wanted = key.toLowerCase()
  let at = block.body.from

  while (at < block.body.to) {
    const end = source.indexOf('\n', at)
    const stop = end === -1 || end > block.body.to ? block.body.to : end
    const found = KEY.exec(source.slice(at, stop))

    if (found?.[1]?.toLowerCase() === wanted) {
      return { from: at, to: stop, value: { from: at + found[0].length, to: stop } }
    }

    if (end === -1) break
    at = end + 1
  }

  return null
}

/** One edit that sets several top-level keys at once, in the order given, or null
 *  where the note already says all of it.
 *
 *  One edit and not several, because the note is written once: two edits into the
 *  same block would have to be given in the coordinates of the text before either of
 *  them, and the second key's line moves when the first one is inserted. So the keys
 *  are set in turn and what comes back is the one span that changed - which for a
 *  note is the block at the top of it, leaving every caret in the words below where
 *  its reader left it.
 *
 *  What two keys at once is for: an icon and the colour it is drawn in. They are two
 *  keys rather than one value so that another app reading the note still finds the
 *  icon, and one write rather than two so that choosing an icon is one thing to undo. */
export function frontMatterEdits(
  source: string,
  keys: readonly (readonly [string, string | null])[],
): TextEdit | null {
  let after = source

  for (const [key, value] of keys) {
    const edit = frontMatterEdit(after, key, value)
    if (edit) after = after.slice(0, edit.from) + edit.insert + after.slice(edit.to)
  }

  return oneEdit(source, after)
}

/** One edit that sets a top-level key to `value`, or takes the key away when it
 *  is null. Null when the note already says that, so a caller writes no file.
 *
 *  What it does in each of the cases a note can be in:
 *
 *  - the key is there: the value after the colon is replaced, and the rest of
 *    the line, the rest of the block and every other key stay as they were.
 *  - the block is there without the key: the key goes in as the last line of the
 *    block, in front of the closing fence.
 *  - there is no block: one is opened at the top of the note, which is where
 *    every reader of a markdown file looks for it.
 *  - taking the key away: its line goes. A block that held nothing else goes
 *    with it, along with the blank lines it was standing on, because an empty
 *    pair of fences at the top of a note is litter rather than metadata.
 *
 *  The line ending is whatever the note already uses, so a file written on
 *  Windows does not come back with one line in the other convention. */
export function frontMatterEdit(
  source: string,
  key: string,
  value: string | null,
): { from: number; to: number; insert: string } | null {
  const nl = source.includes('\r\n') ? '\r\n' : '\n'
  const block = frontMatterBlock(source)

  if (!block) {
    if (value === null) return null
    return { from: 0, to: 0, insert: `---${nl}${key}: ${value}${nl}---${nl}` }
  }

  const line = keyLine(source, block, key)

  if (value === null) {
    if (!line) return null

    // What the block would say without this line. Nothing worth keeping means
    // the fences go too, and the blank lines under them with it.
    const rest =
      source.slice(block.body.from, line.from) + source.slice(cutTo(source, line), block.body.to)
    if (!rest.trim()) return { from: block.from, to: blankAfter(source, block.to), insert: '' }

    return { from: line.from, to: cutTo(source, line), insert: '' }
  }

  if (line) {
    const said = source.slice(line.value.from, line.value.to).trim()
    if (unquoted(said) === value) return null

    return { from: line.value.from, to: line.value.to, insert: ` ${value}` }
  }

  return { from: block.close, to: block.close, insert: `${key}: ${value}${nl}` }
}

/** Where the cut of a key's line ends: past its line break, so the line is gone
 *  rather than left behind as an empty one. */
function cutTo(source: string, line: KeyLine): number {
  return source.startsWith('\r\n', line.to) ? line.to + 2 : Math.min(line.to + 1, source.length)
}

/** Past the empty lines that follow an offset. What a note that opened with
 *  nothing but a block of metadata should start with is its first real line. */
function blankAfter(source: string, at: number): number {
  let over = at
  for (;;) {
    if (source.startsWith('\r\n', over)) over += 2
    else if (source.startsWith('\n', over)) over += 1
    else return over
  }
}

File: apps/desktop/src-tauri/src/front_matter.rs
//! The block of metadata a note may open with, and the two ways one key of it is
//! read.
//!
//! YAML between two `---` fences, which is Obsidian's convention and so this
//! app's: a note that says what it is says it there, and every other editor that
//! reads markdown leaves the block alone.
//!
//! Here rather than in one of the two modules that ask, because both do: `links`
//! reads the icon a row wears and the other names a note gives itself, and `tags`
//! reads the `tags:` key and then leaves the block out of its own pass. A second
//! copy of where the block is would be a second answer to what a note's metadata
//! is.
//!
//! Top-level keys only. A key indented under another one belongs to that one:
//! `paper` under `export:` is not the note's paper, and nothing here pretends
//! otherwise.
//!
//! The twin of this is `packages/markdown/src/front-matter.ts`, which the browser
//! build reads a space with, and the tests at the bottom are the twin of
//! `front-matter.test.ts`: the same notes, the same answers. One difference, which
//! no test on either side fixes: an opening fence indented by a space opens a
//! block here, where the twin asks for it at the left margin.

/// Where a note's front matter sits, in bytes.
///
/// Three offsets, because the two readers ask different things: what the block
/// says is `from` to `close`, and where the note's own words start is `end`.
pub struct Block {
    /// Where the lines between the fences start: just past the opening one.
    pub from: usize,
    /// Where the closing fence's own line begins, which is where those lines end.
    pub close: usize,
    /// Where that line ends, its break not included: the last byte of the block.
    pub end: usize,
}

/// Where the block sits, or None where the note opens with anything else, which
/// is most notes.
///
/// The block has to close, or the note opens with a rule rather than metadata.
pub fn block(body: &str) -> Option<Block> {
    let first = body.find('\n')?;
    if body[..first].trim() != "---" {
        return None;
    }

    let mut at = first + 1;
    while at <= body.len() {
        let end = body[at..].find('\n').map_or(body.len(), |one| at + one);
        if body[at..end].trim() == "---" {
            return Some(Block {
                from: first + 1,
                close: at,
                end,
            });
        }
        if end >= body.len() {
            break;
        }
        at = end + 1;
    }

    None
}

/// A top-level `key: value`, quotes stripped, or None where the note has no
/// block, no such key, or nothing after the colon.
///
/// The twin of `frontMatterValue`.
pub fn value(body: &str, key: &str) -> Option<String> {
    let block = block(body)?;
    lines_of(body, &block).find_map(|line| value_of(line, key))
}

/// A top-level key read as a list: `key: [a, b]`, the `- a` lines written under
/// `key:`, or a single value standing for a list of one. Empty where the note has
/// no block, no such key, or nothing under it.
///
/// The twin of `frontMatterList`.
pub fn list(body: &str, key: &str) -> Vec<String> {
    let Some(block) = block(body) else {
        return Vec::new();
    };

    let mut lines = lines_of(body, &block);
    while let Some(line) = lines.next() {
        let Some(said) = named(line, key) else {
            continue;
        };

        // A key with nothing after it is a list written underneath it.
        return if said.is_empty() {
            dash_items(lines)
        } else {
            flow_items(said)
        };
    }

    Vec::new()
}

/// The block's own lines, without their breaks: from just past the opening fence
/// to the line the closing one is on.
fn lines_of<'a>(body: &'a str, block: &Block) -> std::str::Lines<'a> {
    body.get(block.from..block.close).unwrap_or_default().lines()
}

/// What one line says under `key`, trimmed, or None when the line is another key,
/// a list item, or a line indented under something else.
fn named<'a>(line: &'a str, key: &str) -> Option<&'a str> {
    let (name, said) = line.split_once(':')?;
    // A key of the note's own stands at the left margin; an indented one belongs
    // to the key above it.
    if name.starts_with(char::is_whitespace) || !name.trim_end().eq_ignore_ascii_case(key) {
        return None;
    }

    Some(said.trim())
}

/// What one line says under `key`, or None when it says nothing at all.
fn value_of(line: &str, key: &str) -> Option<String> {
    unquoted(named(line, key)?)
}

/// `[One, Two]` on the key's own line, or a single value standing for a list of
/// one.
fn flow_items(value: &str) -> Vec<String> {
    if let Some(list) = value
        .strip_prefix('[')
        .and_then(|one| one.strip_suffix(']'))
    {
        return list
            .split(',')
            .filter_map(|one| unquoted(one.trim()))
            .collect();
    }

    unquoted(value).into_iter().collect()
}

/// The `- item` lines written under a key, up to the next key of the note's own
/// or anything else that is not an item.
fn dash_items(lines: std::str::Lines<'_>) -> Vec<String> {
    let mut out = Vec::new();

    for line in lines {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let Some(item) = line.strip_prefix('-') else {
            break;
        };
        // A dash with no space after it is not an item, and neither is a dash on
        // its own: `-One` is a word and `-` is a rule.
        if !item.starts_with(char::is_whitespace) {
            break;
        }
        if let Some(said) = unquoted(item.trim()) {
            out.push(said);
        }
    }

    out
}

/// A value with the quotes YAML would take off taken off, or None where nothing
/// is left of it.
fn unquoted(value: &str) -> Option<String> {
    let bare = value
        .strip_prefix('"')
        .and_then(|one| one.strip_suffix('"'))
        .or_else(|| {
            value
                .strip_prefix('\'')
                .and_then(|one| one.strip_suffix('\''))
        })
        .unwrap_or(value);

    if bare.is_empty() {
        None
    } else {
        Some(bare.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::{block, list, value};

    /// The icon a row wears, which is what `links` reads a note's front matter
    /// for. Every case here has its twin in
    /// `packages/markdown/src/front-matter.test.ts`.
    fn icon(body: &str) -> Option<String> {
        value(body, "icon")
    }

    /// The other names a note gives itself, and the other twin.
    fn aliases(body: &str) -> Vec<String> {
        list(body, "aliases")
    }

    #[test]
    fn the_block_is_the_lines_between_two_fences() {
        let note = "---\nicon: rocket\n---\n\n# Plan\n";
        let found = block(note).expect("a block");

        assert_eq!(&note[found.from..found.close], "icon: rocket\n");
        assert_eq!(&note[found.end..], "\n\n# Plan\n");
    }

    #[test]
    fn a_block_with_nothing_in_it_is_still_a_block() {
        let found = block("---\n---\n").expect("a block");

        assert_eq!(found.from, found.close);
        assert_eq!(found.end, 7);
    }

    #[test]
    fn a_note_that_opens_with_anything_else_has_no_block() {
        assert!(block("# Plan\n\nwords").is_none());
        assert!(block("").is_none());
        assert!(block("---").is_none());
        // A block that never closes is a note that opens with a rule.
        assert!(block("---\nicon: rocket\n\n# Plan").is_none());
    }

    #[test]
    fn reads_a_key_from_the_front_matter() {
        assert_eq!(
            icon("---\nicon: rocket\n---\n\n# Plan").as_deref(),
            Some("rocket")
        );
        assert_eq!(
            icon("---\ntitle: Plan\nicon: file-text\n---\n").as_deref(),
            Some("file-text")
        );
        assert_eq!(icon("---\nICON: rocket\n---\n").as_deref(), Some("rocket"));
    }

    #[test]
    fn takes_the_quotes_off_a_value() {
        assert_eq!(
            icon("---\nicon: \"rocket\"\n---\n").as_deref(),
            Some("rocket")
        );
        assert_eq!(
            icon("---\nicon: 'rocket'\n---\n").as_deref(),
            Some("rocket")
        );
    }

    #[test]
    fn reads_an_emoji_as_what_it_says() {
        assert_eq!(icon("---\nicon: 🚀\n---\n").as_deref(), Some("🚀"));
    }

    #[test]
    fn reads_nothing_where_there_is_nothing_to_read() {
        assert_eq!(icon("# Plan\n\nwords"), None);
        assert_eq!(icon("---\ntitle: Plan\n---\n"), None);
        assert_eq!(icon("---\nicon:\n---\n"), None);
        // A block that never closes is a note that opens with a rule.
        assert_eq!(icon("---\nicon: rocket\n\n# Plan"), None);
        // And a key below the block is the note's words rather than its metadata.
        assert_eq!(icon("---\ntitle: Plan\n---\n\nicon: rocket\n"), None);
    }

    #[test]
    fn a_key_indented_under_another_is_not_the_notes_own() {
        assert_eq!(icon("---\nexport:\n  icon: rocket\n---\n"), None);
    }

    /// A note written on Windows carries `\r\n`, and the carriage return is no
    /// part of what the key says.
    #[test]
    fn reads_a_note_written_with_carriage_returns() {
        assert_eq!(
            icon("---\r\nicon: rocket\r\n---\r\n").as_deref(),
            Some("rocket")
        );
    }

    #[test]
    fn reads_a_list_written_on_the_key_s_own_line() {
        assert_eq!(
            aliases("---\naliases: [One, Two]\n---\n"),
            vec!["One".to_string(), "Two".to_string()]
        );
    }

    #[test]
    fn reads_a_list_written_under_the_key() {
        assert_eq!(
            aliases("---\naliases:\n  - One\n  - Two\n---\n"),
            vec!["One".to_string(), "Two".to_string()]
        );
        assert_eq!(
            aliases("---\naliases:\n- One\n---\n"),
            vec!["One".to_string()]
        );
    }

    #[test]
    fn reads_a_single_value_as_a_list_of_one() {
        assert_eq!(aliases("---\naliases: One\n---\n"), vec!["One".to_string()]);
    }

    #[test]
    fn takes_the_quotes_off_every_item() {
        assert_eq!(
            aliases("---\naliases: [\"One\", 'Two']\n---\n"),
            vec!["One".to_string(), "Two".to_string()]
        );
        assert_eq!(
            aliases("---\naliases:\n  - \"One two\"\n---\n"),
            vec!["One two".to_string()]
        );
    }

    #[test]
    fn stops_at_the_next_key_of_the_notes_own() {
        assert_eq!(
            aliases("---\naliases:\n  - One\ntitle: Not an alias\n---\n"),
            vec!["One".to_string()]
        );
    }

    #[test]
    fn reads_nothing_where_there_is_no_list() {
        assert!(aliases("Words.\n").is_empty());
        assert!(aliases("---\ntitle: A note\n---\n").is_empty());
        assert!(aliases("---\naliases:\n---\n").is_empty());
        assert!(aliases("---\naliases: []\n---\n").is_empty());
        assert!(aliases("---\nexport:\n  aliases: [One]\n---\n").is_empty());
    }
}
